File: ops/pypto/torch_ops_extension/custom_pypto/converter/npu_sparse_attention_pto.py
# This program is free software, you can redistribute it and/or modify it.
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    Any, Callable, ContextManager, Iterator, List, Literal, NamedTuple, Optional, Sequence, Tuple, TypeVar,
    Union, overload,
)
import torch
import torch_npu
import torchair
from torch import Generator, contiguous_format, inf, strided, SymInt
from torch.types import Device, Number, _bool, _complex, _device, _dtype, _float, _int, _layout, _qscheme, _size
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import declare_supported, register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec, DataType
from torchair._ge_concrete_graph.supported_declaration import _TypedTensor, F32, F16, F64, I32, I16, I64, I8, U8, \
    BOOL, Support
from torchair._ge_concrete_graph.utils import dtype_promote
from torchair.ge import attr


# 为自定义算子注册converter，用于torch.compile 场景成图

# 注意： meta_outputs形参名为固定写法，若写错会影响ge节点的输出dtype与shape推导
@register_fx_node_ge_converter(torch.ops.custom_pypto.npu_sparse_attention_pto.default)
def convert_npu_sparse_attention_pto(
    x: Tensor,
    w_dq: Tensor,
    w_uq_qr: Tensor,
    w_uk: Tensor,
    w_dkv_kr: Tensor,
    gamma_cq: Tensor,
    gamma_ckv: Tensor,
    sin: Tensor,
    cos: Tensor,
    cache_index: Tensor,
    kv_cache: Tensor,
    kr_cache: Tensor,
    block_table: Tensor,
    act_seqs: Tensor,
    w_idx_qb: Tensor,
    w_idx_k: Tensor,
    w_idx_proj: Tensor,
    in_gamma_k: Tensor,
    in_beta_k: Tensor,
    index_k_cache: Tensor,
    meta_outputs: Any = None
    ):

    '''NB: npu_sparse_attention_pto(Tensor x, Tensor w_dq, Tensor w_uq_qr, Tensor w_uk, Tensor w_dkv_kr, Tensor gamma_cq, Tensor gamma_ckv,
                                    Tensor sin, Tensor cos, Tensor cache_index, Tensor kv_cache, Tensor kr_cache, Tensor block_table,
                                    Tensor act_seqs, Tensor w_idx_qb, Tensor w_idx_k, Tensor w_idx_proj, Tensor in_gamma_k, Tensor in_beta_k, Tensor index_k_cache) -> Tensor'''

    input_list = [
        x, w_dq, w_uq_qr, w_uk, w_dkv_kr, gamma_cq, gamma_ckv,
        sin, cos, cache_index, kv_cache, kr_cache, block_table,
        act_seqs, w_idx_qb, w_idx_k, w_idx_proj, in_gamma_k, in_beta_k, index_k_cache
    ]
    
    return torchair.ge.custom_op(
        "SparseAttentionPto",
        inputs={
            "x0": x,
            "x1": w_dq,
            "x2": w_uq_qr,
            "x3": w_uk,
            "x4": w_dkv_kr,
            "x5": gamma_cq,
            "x6": gamma_ckv,
            "x7": sin,
            "x8": cos,
            "x9": cache_index,
            "x10": kv_cache,
            "x11": kr_cache,
            "x12": block_table,
            "x13": act_seqs,
            "x14": w_idx_qb,
            "x15": w_idx_k,
            "x16": w_idx_proj,
            "x17": in_gamma_k,
            "x18": in_beta_k,
            "x19": index_k_cache,
        },
        outputs=['y0']
    )

File: ops/pypto/torch_ops_extension/custom_pypto/csrc/npu_sparse_attention_pto.cpp
/* *
 * This program is free software, you can redistribute it and/or modify it.
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.  * See LICENSE in the root of
the software repository for the full text of the License.
 */

#include <iostream>
#include <torch/library.h>
#include "torch_npu/csrc/aten/CustomFunctions.h"
#include "torch_npu/csrc/aten/NPUNativeFunctions.h"
#include "torch_npu/csrc/core/npu/NPUStream.h"
#include "torch_npu/csrc/core/npu/NPUFunctions.h"
#include "torch_npu/csrc/core/npu/NpuVariables.h"
#include "torch_npu/csrc/core/npu/register/OptionsManager.h"
#include "torch_npu/csrc/framework/OpCommand.h"
#include <torch_npu/csrc/framework/utils/CalcuOpUtil.h>
#include <torch_npu/csrc/framework/utils/OpAdapter.h>
#include "torch_npu/csrc/framework/utils/OpPreparation.h"
#include "torch_npu/csrc/framework/utils/RandomOpAdapter.h"
#include "torch_npu/csrc/framework/interface/AclOpCompileInterface.h"
#include "torch_npu/csrc/framework/interface/EnvVariables.h"
#include "torch_npu/csrc/flopcount/FlopCount.h"
#include "torch_npu/csrc/flopcount/FlopCounter.h"
#include "torch_npu/csrc/custom_dtype/Init.h"

namespace custom_pypto {
using namespace at_npu::native;

// npu tensor max size
const int SIZE = 8;
const int DIM_0 = 0;
const int DIM_1 = 1;
const int DIM_2 = 2;
const int DIM_3 = 3;

// 工具函数，推导输出shape
at::Tensor construct_npu_sparse_attention_pto_output_tensor(
    const at::Tensor &x, const at::Tensor &w_dq, const at::Tensor &w_uq_qr, const at::Tensor &w_uk,
    const at::Tensor &w_dkv_kr, const at::Tensor &gamma_cq, const at::Tensor &gamma_ckv,
    const at::Tensor &sin, const at::Tensor &cos, const at::Tensor &cache_index,
    const at::Tensor &kv_cache, const at::Tensor &kr_cache, const at::Tensor &block_table, const at::Tensor &act_seqs,
    const at::Tensor &w_idx_qb, const at::Tensor &w_idx_k, const at::Tensor &w_idx_proj,
    const at::Tensor &in_gamma_k,const at::Tensor &in_beta_k, const at::Tensor &index_k_cache
)
{
    at::SmallVector<int64_t, SIZE> output_size;
    // BSND
    output_size = {x.size(DIM_0), x.size(DIM_1), w_uk.size(DIM_0), w_uk.size(DIM_2)};
    at::Tensor output = at::empty(output_size, x.options().dtype(x.dtype()));

    return output;
}

// step3, 为META设备实现前向接口
at::Tensor npu_sparse_attention_pto_meta(
    const at::Tensor &x, const at::Tensor &w_dq, const at::Tensor &w_uq_qr, const at::Tensor &w_uk,
    const at::Tensor &w_dkv_kr, const at::Tensor &gamma_cq, const at::Tensor &gamma_ckv,
    const at::Tensor &sin, const at::Tensor &cos, const at::Tensor &cache_index,
    const at::Tensor &kv_cache, const at::Tensor &kr_cache, const at::Tensor &block_table, const at::Tensor &act_seqs,
    const at::Tensor &w_idx_qb, const at::Tensor &w_idx_k, const at::Tensor &w_idx_proj,
    const at::Tensor &in_gamma_k,const at::Tensor &in_beta_k, const at::Tensor &index_k_cache
)
{
    // input_list = [query, key, weights, actual_seq_lengths_key, block_table]
    return construct_npu_sparse_attention_pto_output_tensor(
            x, w_dq, w_uq_qr, w_uk, w_dkv_kr, gamma_cq, gamma_ckv,
            sin, cos, cache_index, kv_cache, kr_cache, block_table, act_seqs,
            w_idx_qb, w_idx_k, w_idx_proj, in_gamma_k, in_beta_k, index_k_cache
    );
}
}

// step5, 为META设备注册前向实现
TORCH_LIBRARY_IMPL(custom_pypto, Meta, m) {
    m.impl("npu_sparse_attention_pto", &custom_pypto::npu_sparse_attention_pto_meta);
}
